File: cs/etsm/test/ab.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Etsm;

namespace ab
{
    public class Foo
    {
        private StateMachine<State> stateMachine;
        private State stateA;
        private State stateB;

        private int enterStateA = 0;
        private int exitStateA = 0;
        private int enterStateB = 0;
        private int exitStateB = 0;

        public Foo()
        {
            stateMachine = new StateMachine<State>();
            stateA = new State(EnterA, ExitA);
            stateB = new State(EnterB, ExitB);
        }

        // Simple simulation that make transition using this pattern:
        //    None -> A -> B -> None
        public void Simulation()
        {
            if (stateMachine.IsIn(null))
                stateMachine.Transition(stateA);
            else if (stateMachine.IsIn(stateA))
                stateMachine.Transition(stateB);
            else if (stateMachine.IsIn(stateB))
                stateMachine.Transition(null);
        }

        void EnterA()
        {
            enterStateA++;
        }

        void ExitA()
        {
            exitStateA++;
        }

        void EnterB()
        {
            enterStateB++;
        }

        void ExitB()
        {
            exitStateB++;
        }

        public void Test()
        {
            for (int i = 1; i < 3; i++)
            {
                Assert.IsTrue(stateMachine.IsIn(null));

                Simulation();
                Assert.IsTrue(stateMachine.IsIn(stateA));

                Simulation();
                Assert.IsTrue(stateMachine.IsIn(stateB));

                Simulation();
                Assert.IsTrue(stateMachine.IsIn(null));

                Assert.AreEqual(i, enterStateA);
                Assert.AreEqual(i, exitStateA);
                Assert.AreEqual(i, enterStateB);
                Assert.AreEqual(i, exitStateB);
            }
        }
    }

    [TestClass]
    public class ab
    {
        [TestMethod]
        public void Test()
        {
            Foo foo = new Foo();
            foo.Test();
        }
    }
}

File: rust/etsm/tests/ab.rs
use etsm::*;

state_machine!(
    Foo,
    State {
        {A, Some(Foo::enter_a), Some(Foo::exit_a)},
        {B, Some(Foo::enter_b), Some(Foo::exit_b)}
    }
);

struct Foo {
    state_machine: StateMachine<State>,
    state_a_enter: i32,
    state_a_exit: i32,
    state_b_enter: i32,
    state_b_exit: i32,
}

impl Foo {
    fn new() -> Foo {
        Foo {
            state_machine: StateMachine::new(),
            state_a_enter: 0,
            state_a_exit: 0,
            state_b_enter: 0,
            state_b_exit: 0,
        }
    }

    // Simple simulation that make transition using this pattern:
    //    None -> A -> B -> None
    fn simulation(&mut self) {
        match &self.state_machine.current_state {
            None => {
                transition!(self, state_machine, Some(State::A));
            }
            Some(s) => match s {
                State::A => {
                    transition!(self, state_machine, Some(State::B));
                }
                State::B => {
                    transition!(self, state_machine, None);
                }
            },
        }
    }

    fn enter_a(&mut self) {
        self.state_a_enter += 1;
    }

    fn exit_a(&mut self) {
        self.state_a_exit += 1;
    }

    fn enter_b(&mut self) {
        self.state_b_enter += 1;
    }

    fn exit_b(&mut self) {
        self.state_b_exit += 1;
    }
}

#[test]
fn ab() {
    let mut foo = Foo::new();

    for i in 1..3 {
        assert!(foo.state_machine.is_in(None));
        foo.simulation();
        assert!(foo.state_machine.is_in(Some(State::A)));
        foo.simulation();
        assert!(foo.state_machine.is_in(Some(State::B)));
        foo.simulation();
        assert!(foo.state_machine.is_in(None));
        assert_eq!(foo.state_a_enter, i);
        assert_eq!(foo.state_a_exit, i);
        assert_eq!(foo.state_b_enter, i);
        assert_eq!(foo.state_b_exit, i);
    }
}
